inspectkv/inspectkv.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/terror"
+)
+
+// DDLInfo is for DDL information.
+type DDLInfo struct {
+	SchemaVer   int64
+	ReorgHandle int64
+	Owner       *model.Owner
+	Job         *model.Job
+}
+
+// GetDDLInfo returns DDL information.
+func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
+	var err error
+	info := &DDLInfo{}
+	t := meta.NewMeta(txn)
+
+	info.Owner, err = t.GetDDLOwner()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info.Job, err = t.GetDDLJob(0)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info.SchemaVer, err = t.GetSchemaVersion()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if info.Job == nil {
+		return info, nil
+	}
+
+	info.ReorgHandle, err = t.GetDDLReorgHandle(info.Job)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return info, nil
+}
+
+// GetIndexVals returns index values.
+func GetIndexVals(table table.Table, txn kv.Transaction, indexInfo model.IndexInfo, handle int64) ([]interface{}, error) {
+	vals := make([]interface{}, len(indexInfo.Columns))
+	cols := table.Cols()
+
+	for i, col := range indexInfo.Columns {
+		key := table.RecordKey(handle, cols[col.Offset])
+		val, err := txn.Get(key)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		vals[i] = val
+	}
+
+	return vals, nil
+}
+
+// GetIndexHandles returns index handles.
+func GetIndexHandles(table table.Table, txn kv.Transaction, idx *column.IndexedCol) ([]int64, error) {
+	var handles []int64
+	kvIndex := kv.NewKVIndex(table.IndexPrefix(), idx.Name.L, idx.ID, idx.Unique)
+	it, err := kvIndex.SeekFirst(txn)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for {
+		_, h, err := it.Next()
+		if terror.ErrorEqual(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, errors.Trace(err)
+		}
+		handles = append(handles, h)
+	}
+
+	return handles, nil
+}
+
+// GetTableData gets table row handles and column values.
+// If there is no special iterator, it will be nil.
+func GetTableData(table table.Table, txn kv.Transaction, it kv.Iterator) ([]int64, []interface{}, error) {
+	var handles []int64
+	var data []interface{}
+
+	err := table.IterRecords(txn, it, table.FirstKey(), table.Cols(),
+		func(h int64, d []interface{}, cols []*column.Col) (bool, error) {
+			data = append(data, d)
+			handles = append(handles, h)
+			return true, nil
+		})
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	return handles, data, nil
+}
+
+// GetTableSnapshot gets the ver version of the table data.
+func GetTableSnapshot(store kv.Storage, ver kv.Version, table table.Table, txn kv.Transaction) ([]interface{}, error) {
+	snap, err := store.GetSnapshot(ver)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer snap.Release()
+
+	it, err := snap.Seek([]byte(table.FirstKey()))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer it.Close()
+
+	_, data, err := GetTableData(table, txn, it)
```

Comment:
the snapshot for the txn is not the same for store.GetSnapshot(ver), why using them together?